reassembly/tcpassembly.go

```diff
@@ -0,0 +1,1227 @@
+// run at once to assemble packets while taking advantage of multiple cores.
+package reassembly
+
+import (
+	"flag"
+	"fmt"
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+	"log"
+	"sync"
+	"time"
+)
+
+// TODO:
+// - push to Stream on Ack
+// - implement chunked (cheap) reads and Reader() interface
+
+var defaultDebug = false
+
+var debugLog = flag.Bool("assembly_debug_log", defaultDebug, "If true, the github.com/google/gopacket/reassembly library will log verbose debugging information (at least one line per packet)")
+
+const invalidSequence = -1
+const uint32Max = 0xFFFFFFFF
+
+// Sequence is a TCP sequence number.  It provides a few convenience functions
+// for handling TCP wrap-around.  The sequence should always be in the range
+// [0,0xFFFFFFFF]... its other bits are simply used in wrap-around calculations
+// and should never be set.
+type Sequence int64
+
+// Difference defines an ordering for comparing TCP sequences that's safe for
+// roll-overs.  It returns:
+//    > 0 : if t comes after s
+//    < 0 : if t comes before s
+//      0 : if t == s
+// The number returned is the sequence difference, so 4.Difference(8) will
+// return 4.
+//
+// It handles rollovers by considering any sequence in the first quarter of the
+// uint32 space to be after any sequence in the last quarter of that space, thus
+// wrapping the uint32 space.
+func (s Sequence) Difference(t Sequence) int {
+	if s > uint32Max-uint32Max/4 && t < uint32Max/4 {
+		t += uint32Max
+	} else if t > uint32Max-uint32Max/4 && s < uint32Max/4 {
+		s += uint32Max
+	}
+	return int(t - s)
+}
+
+// Add adds an integer to a sequence and returns the resulting sequence.
+func (s Sequence) Add(t int) Sequence {
+	return (s + Sequence(t)) & uint32Max
+}
+
+type TCPAssemblyStats struct {
+	// For this ScatterGather
+	NbChunks  uint
+	NbPackets uint
+	// For the half connection, since last call to ReassembledSG()
+	NbQueuedBytes    uint
+	NbQueuedPackets  uint
+	NbOverlapBytes   uint
+	NbOverlapPackets uint
+}
+
+type ScatterGather interface {
+	// Returns the length of available bytes and saved bytes
+	Lengths() (int, int)
+	// Returns the bytes up to length
+	Fetch(length int) []byte
+	// Tell to keep from offset
+	KeepFrom(offset int)
+	// Return CaptureInfo of packet corresponding to given offset
+	CaptureInfo(offset int) gopacket.CaptureInfo
+	// Return some info about the reassembled chunks
+	Info() (direction TcpFlowDirection, start bool, end bool, skip int)
+	// Return some stats regarding the state of the stream
+	Stats() TCPAssemblyStats
+}
+
+// Containter: either page or LivePacket
+type ByteContainer interface {
+	Bytes() []byte
+	Length() int
+	ConvertToPages(*pageCache, int, AssemblerContext) (*page, *page, int)
+	CaptureInfo() gopacket.CaptureInfo
+	AssemblerContext() AssemblerContext
+	Release(*pageCache) int
+	Start() bool
+	End() bool
+	Seq() Sequence
+	IsPacket() bool
+}
+
+// Implements a ScatterGather
+type ReassemblyObject struct {
```

Comment:
should this be exported?  If it's being passed as a ScatterGather, there doesn't seem to be a need.